### PR TITLE
Dev deviceprovisioningservices microsoft.devices 2021 02 01 preview

### PR DESCRIPTION
**Pull request by sudo1997lsl**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [x] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  Target date to deploy new feature: 03/22/2021
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [x] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json [hunk 1]
```diff
@@ -0,0 +1,2619 @@
...
+        "keyName": {
+          "type": "string",
+          "description": "Name of the key."
+        },
+        "primaryKey": {
+          "type": "string",
+          "description": "Primary SAS key value."
+        },
+        "secondaryKey": {
+          "type": "string",
```
Comment (j5lim) [inline specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json:2243]: Please annotate with x-ms-secret. All secrets should be marked as "x-ms-secret": "true"

File: specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json [hunk 2]
```diff
@@ -0,0 +1,2619 @@
...
+            "modelAsString": true
+          }
+        }
+      },
+      "required": [
+        "keyName",
+        "rights"
+      ]
+    },
+    "Resource": {
```
Comment (j5lim) [inline specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json:2268]: In a new api version, Azure resource should support a top level property `systemData`.

https://armwiki.azurewebsites.net/api_contracts/ResourceSystemData.html

You can reference the one in common-type.
https://github.com/Azure/azure-rest-api-specs/blob/a16488bbd0d17551c729e22ea6ac6f7c6588a0a3/specification/common-types/resource-management/v2/types.json#L445

File: specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json [hunk 3]
```diff
@@ -0,0 +1,2619 @@
...
+        },
+        "type": {
+          "readOnly": true,
+          "type": "string",
+          "description": "The resource type."
+        },
+        "properties": {
+          "type": "object",
+          "properties": {
+            "verificationCode": {
```
Comment (j5lim) [inline specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json:2527]: Is this a secret? Then please annotate with x-ms-secret. GET. https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html#common-issues-found-in-review

**Review by j5lim — CHANGES_REQUESTED** [review]

File: specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json (hunk 3 above)
Comment (sudo1997lsl) [reply specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json:2527]: @mrohera Can you confirm that this is a secret?

Comment (sudo1997lsl) [reply specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json:2529]: x-ms-secret: true added

File: specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json [hunk 4]
```diff
@@ -0,0 +1,2622 @@
...
+            "type": "string",
+            "maxLength": 256
+          },
+          {
+            "name": "certificateDescription",
+            "in": "body",
+            "description": "The certificate body.",
+            "required": true,
+            "schema": {
+              "$ref": "#/definitions/CertificateBodyDescription"
```
Comment (j5lim) [inline specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json:190]: Please use the same definition for both the request body and response. You can mark properties in the response only as read-only.

File: specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json [hunk 5]
```diff
@@ -0,0 +1,2622 @@
...
+          },
+          {
+            "name": "certificateName",
+            "in": "path",
+            "required": true,
+            "type": "string",
+            "description": "This is a mandatory field, and is the logical name of the certificate that the provisioning service will access by."
+          },
+          {
+            "name": "certificate.name",
```
Comment (j5lim) [inline specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json:272]: Are these query strings used for DELETE?

File: specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/examples/DPSCreateOrUpdatePrivateEndpointConnection.json [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        }
+      }
+    },
+    "201": {
+      "body": {
+        "id": "/subscriptions/91d12660-3dec-467a-be2a-213b5544ddc0/resourceGroups/myResourceGroup/providers/Microsoft.Devices/ProvisioningServices/myFirstProvisioningService/PrivateEndpointConnections/myPrivateEndpointConnection",
+        "name": "myPrivateEndpointConnection",
+        "type": "Microsoft.Devices/ProvisioningServices/PrivateEndpointConnections",
+        "properties": {
+          "privateEndpoint": {
```
Comment (j5lim) [inline specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/examples/DPSCreateOrUpdatePrivateEndpointConnection.json:41]: Please add provisiongState to indicate the state of the async operation.
https://armwiki.azurewebsites.net/api_contracts/guidelines/rpc.html#rpc005-provisioning-state-semantics-must-be-followed

File: specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json [hunk 7]
```diff
@@ -0,0 +1,2622 @@
...
+            "required": true,
+            "type": "string",
+            "description": "Resource group identifier."
+          },
+          {
+            "$ref": "#/parameters/api-version"
+          }
+        ],
+        "responses": {
+          "202": {
```
Comment (j5lim) [inline specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json:584]: Please consider adding "headers" property to specify Location and Azure-AsyncOperation header in the response and also update the example to show the headers there as well.

File: specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json [hunk 8]
```diff
@@ -0,0 +1,2622 @@
...
+          "default": {
+            "description": "Default error response.",
+            "schema": {
+              "$ref": "#/definitions/ErrorDetails"
+            }
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Devices/provisioningServices/{provisioningServiceName}/certificates": {
```
Comment (j5lim) [inline specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json:605]: Please add x-ms-pageable to all List apis even if paging is not supported.
https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html?q=x-ms-pageable#oapi009-always-add-x-ms-pageable-to-list-calls

File: specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json [hunk 9]
```diff
@@ -0,0 +1,2622 @@
...
+          },
+          {
+            "name": "provisioningServiceName",
+            "in": "path",
+            "required": true,
+            "type": "string",
+            "description": "Name of provisioning service that the operation is running on."
+          },
+          {
+            "name": "asyncinfo",
```
Comment (j5lim) [inline specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json:791]: Why is this needed? The GET operationsresult URL should be sufficient to get the async operation state.

File: specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/examples/DPSGetOperationResult.json [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+    "provisioningServiceName": "myFirstProvisioningService",
+    "resourceGroupName": "myResourceGroup",
+    "api-version": "2021-02-01-preview",
+    "subscriptionId": "91d12660-3dec-467a-be2a-213b5544ddc0",
+    "operationId": "MTY5OTNmZDctODI5Yy00N2E2LTkxNDQtMDU1NGIyYzY1ZjRl",
+    "asyncinfo": "1508265712453"
+  },
+  "responses": {
+    "200": {
+      "body": {
```
Comment (j5lim) [inline specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/examples/DPSGetOperationResult.json:12]: If this is the response from the URI specified in the Location header, it needs to return the exactly same response as that would have been returned had the operation been executed synchronously.

https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#202-accepted-and-location-headers

File: specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json [hunk 11]
```diff
@@ -0,0 +1,2622 @@
...
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Devices/provisioningServices/{provisioningServiceName}/certificates/{certificateName}/generateVerificationCode": {
+      "post": {
+        "tags": [
+          "POST"
+        ],
+        "operationId": "DpsCertificate_GenerateVerificationCode",
+        "description": "Generate verification code for Proof of Possession.",
+        "x-ms-examples": {
+          "DPSGenerateVerificationCode": {
+            "$ref": "./examples/DPSGenerateVerificationCode.json"
```
Comment (j5lim) [inline specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json:881]: Adding more examples with the supported query strings will be helpful for customers.

**Review by j5lim — CHANGES_REQUESTED** [review]

**Comment by raych1** [comment]
@sudo1997lsl , regard to the `secret_property` error in model validation, this is by design. Please suppress this error in `readme.md` in case of `post` scenario like below.
## Suppression
``` yaml
directive:
  - suppress: SECRET_PROPERTY
    from: xxx.json
    where:
      - $.definitions.xxx
```

**Comment by sudo1997lsl** [comment]
> @sudo1997lsl , regard to the `secret_property` error in model validation, this is by design. Please suppress this error in `readme.md` in case of `post` scenario like below.
> 
> ## Suppression
> ```yaml
> directive:
>   - suppress: SECRET_PROPERTY
>     from: xxx.json
>     where:
>       - $.definitions.xxx
> ```

File: specification/deviceprovisioningservices/resource-manager/readme.md [hunk 12]
```diff
@@ -105,6 +114,19 @@ swagger-to-sdk:
...
 ```
 
+## Suppression
+``` yaml
+directive:
+  - suppress: SECRET_PROPERTY
+    from: 
+      - Microsoft.Devices/preview/2021-02-01-preview/iotdps.json
+    where: 
+      - $.definitions.SharedAccessSignatureAuthorizationRule[AccessRightsDescription].properties.primaryKey
```
Comment (raych1) [inline specification/deviceprovisioningservices/resource-manager/readme.md:124]: Square bracket is not supported in property name when do suppression.

File: specification/deviceprovisioningservices/resource-manager/readme.md [hunk 13]
```diff
@@ -105,6 +114,19 @@ swagger-to-sdk:
...
+      - $.definitions.SharedAccessSignatureAuthorizationRule[AccessRightsDescription].properties.secondaryKey
+      - $.definitions.VerificationCodeResponse.properties.verificationCode
```
Comment (raych1) [inline specification/deviceprovisioningservices/resource-manager/readme.md:126]: The correct property name is $.definitions.VerificationCodeResponse.properties.properties.properties.verificationCode

**Comment by raych1** [comment]
@akning-ms , model validation failures are false alarms and I created a [task](https://devdiv.visualstudio.com/DevDiv/_queries/edit/1267074/?triage=true) to improve the tool to ignore secret check in case of post operations. Please force merge this PR for now, thanks.

File: specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json (hunk 4 above)
Comment (sudo1997lsl) [reply specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json:190]: After discussion with @j5lim in the ARM review meeting on Jan 14. We agreed to temporarily resolves these conversations as they are not related to the changes introduced in this 2021-02-01-preview version. Also dps team needs to prioritize this new preview API and deliver it in a timely manner

Comments in this PR are captured in these 2 items and will get fixed in the next swagger PR in the future.
Product Backlog Item [9134544](https://msazure.visualstudio.com/One/_workitems/edit/9134544): Resolve Comments in RP 20210201Preview Swagger PR
Product Backlog Item [9134490](https://msazure.visualstudio.com/One/_workitems/edit/9134490): Add systemData Property in RP Swagger

File: specification/deviceprovisioningservices/resource-manager/readme.md (hunk 13 above)
Comment (sudo1997lsl) [reply specification/deviceprovisioningservices/resource-manager/readme.md]: Fixed

File: specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json (hunk 5 above)
Comment (sudo1997lsl) [reply specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json:272]: Temporarily resolved. See details in this [comment](https://github.com/Azure/azure-rest-api-specs/pull/12378#discussion_r560402312)

File: specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/examples/DPSCreateOrUpdatePrivateEndpointConnection.json (hunk 6 above)
Comment (sudo1997lsl) [reply specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/examples/DPSCreateOrUpdatePrivateEndpointConnection.json:41]: Temporarily resolved. See details in this [comment](https://github.com/Azure/azure-rest-api-specs/pull/12378#discussion_r560402312)

File: specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json (hunk 7 above)
Comment (sudo1997lsl) [reply specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json:584]: Temporarily resolved. See details in this [comment](https://github.com/Azure/azure-rest-api-specs/pull/12378#discussion_r560402312)

File: specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json (hunk 8 above)
Comment (sudo1997lsl) [reply specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json:605]: Temporarily resolved. See details in this [comment](https://github.com/Azure/azure-rest-api-specs/pull/12378#discussion_r560402312)

File: specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json (hunk 9 above)
Comment (sudo1997lsl) [reply specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json:791]: Temporarily resolved. See details in this [comment](https://github.com/Azure/azure-rest-api-specs/pull/12378#discussion_r560402312)

File: specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/examples/DPSGetOperationResult.json (hunk 10 above)
Comment (sudo1997lsl) [reply specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/examples/DPSGetOperationResult.json:12]: Temporarily resolved. See details in this [comment](https://github.com/Azure/azure-rest-api-specs/pull/12378#discussion_r560402312)

File: specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json (hunk 11 above)
Comment (sudo1997lsl) [reply specification/deviceprovisioningservices/resource-manager/Microsoft.Devices/preview/2021-02-01-preview/iotdps.json:881]: Temporarily resolved. See details in this [comment](https://github.com/Azure/azure-rest-api-specs/pull/12378#discussion_r560402312)

File: specification/deviceprovisioningservices/resource-manager/readme.md (hunk 12 above)
Comment (sudo1997lsl) [reply specification/deviceprovisioningservices/resource-manager/readme.md:124]: Will leave the bracket in property name for now and will depend on a force merge for the validation failures caused by this.
Long term fix will be false alarms mentioned [here](https://github.com/Azure/azure-rest-api-specs/pull/12378#issuecomment-761957826)

**Review by j5lim — APPROVED** [review]
Signing off from ARM side.

**Comment by akning-ms** [comment]
seems model validation still not able to suppress. @raych1 
Hi @sudo1997lsl , ping me if you are ok to merge. saw you added label DoNotMerge

**Comment by raych1** [comment]
> seems model validation still not able to suppress. @raych1
> Hi @sudo1997lsl , ping me if you are ok to merge. saw you added label DoNotMerge

yes, as I said the suppression doesn't support square in property name. https://github.com/Azure/azure-rest-api-specs/pull/12378#discussion_r559297052.

**Comment by akning-ms** [comment]
Hi @sudo1997lsl , Can I know what is the status for this PR? I saw it is label  DoNotMerge for a long time.

**Comment by akning-ms** [comment]
> Hi @akning-ms , We are still working on developing the feature to support the changes in this swagger PR. I thought we are not supposed to merge until our service deploy the changes. Is that correct?

yes. thanks for confirm, and can you put your deploy ETA in 2nd item of change log in first comment

**Comment by sudo1997lsl** [comment]
We no longer need the 2021-02-01-preview IoT DPS RP API due to DPS team's design change. I am going to abandon and close this PR.
@akning-ms, let me know if there are any concerns. Thanks

**Comment by sudo1997lsl** [comment]
Close this pull request as it is not needed due to design change
